nixos/modules/services/x11/xserver.nix

```diff
@@ -220,6 +220,11 @@ in
         default = true;
         description = ''
           Whether to start the X server automatically.
+
+          ::: {.note}
+          If false, `systemd.defaultUnit = lib.mkForce "multi-user.target";`
+          may need to be set for this option to take effect.
```

Comment:
I don't know if this makes too much sense to me. Even after reading the linked issue, I don't really get it. 

1. Set `services.xserver.enable = true`
2. Set `services.xserver.autorun = false`
3. See the X server start anyway?

Reply:
@philiptaron Yes that is correct.  When first booting or enabling a configuration that has services.xserver.autorun = false, the xserver runs anyway as if it was set to true.  Manually setting systemd.defaultUnit = lib.mkForce "multi-user.target" will cause the autorun = false setting to work as expected, and it won't run until started manually with "systemctl start display-manager"

Reply:
I see; it's this line that's "at fault" and that this documentation is trying to point at.

https://github.com/NixOS/nixpkgs/blob/bc4d5823e485c6ab32830cf89fceacd5afb1c5d3/nixos/modules/services/misc/graphical-desktop.nix#L64

In particular, it's `dmcfg.enable` or `services.displayManager.enable` that's causing this auto-start behavior. That module is over in [`nixos/modules/services/display-managers/default.nix`](https://github.com/NixOS/nixpkgs/blob/bc4d5823e485c6ab32830cf89fceacd5afb1c5d3/nixos/modules/services/display-managers/default.nix#L36).

Instead of changing the target, is the right thing to do here to guide to the `services.displayManager.enable` configuration also?

Reply:
I believe ```services.displayManager.enable``` is already set to true by default when xserver is enabled.
https://github.com/NixOS/nixpkgs/blob/7c9fb83bf56535f6f6897066b9eece58be7e4c93/nixos/modules/services/x11/xserver.nix#L644
So adding ```services.displayManager.enable = true``` manually to ```configuration.nix``` in itself doesn't have any effect in causing ```services.xserver.autorun = false``` to behave as it should.

Reply:
I guess the root of my confusion is this: what do you / others that set `services.xserver.autorun = false` actually want? A TTY? If that's the case, setting `services.displayManager.enable = lib.mkForce false;` will... likely do it?

I'm sort of stumbling in the dark (as it were) trying to figure out what this combination of options ought to produce.

Reply:
In the past, ```services.xserver.autorun = false``` caused NixOS to boot into a TTY, not straight into the window manager.   To get a graphical desktop the user would have to run ```systemctl start display-mananger```, at which point whatever displayManager+windowManager was configured would start.

We cannot set ```services.displayManager.enable = false``` because it has to be true for xserver to run with its defaults.  Forcing ```services.displayManager.enable``` to false when ```services.xserver.enable``` is true causes ```nixos-rebuild``` to fail when evaluating ```lightdm``` or ```gdm```.

Let me go back in time to find out which commit it was that stopped ```autorun = false``` from working the way it use to and I'll report back.

Reply:
@philiptaron In release-23.11 xserver.nix handled the logic for setting graphic.target based only on the autorun value:
https://github.com/NixOS/nixpkgs/blob/7144d6241f02d171d25fba3edeaf15e0f2592105/nixos/modules/services/x11/xserver.nix#L752
This logic was moved to graphical-desktop.nix in release-24.05 which also checks if displayManager enable is set.  The current logic checks if either is true, but does not take into account if autorun is false.
https://github.com/NixOS/nixpkgs/blob/bde883b11ec0c83cb9c7327e4d12cd189c354fed/nixos/modules/services/misc/graphical-desktop.nix#L64

We could change the logic to only work if autorun is not false with:
```nix
systemd.defaultUnit = lib.mkIf (xcfg.autorun != false && (xcfg.autorun || dmcfg.enable)) "graphical.target";
```
updating graphical-desktop.nix this way makes it so defaultUnit is not set to graphical.target if autorun is set to false.

Let me know if you prefer this approach and I can either open a new PR or submit the change here granted @SuperSandro2000 does not see any conflicts

Reply:
I'm interested in @SuperSandro2000's strenuous objections which produced this documentation patch instead of a patch which changes the semantics.

Reply:
We could do the following which is basically the same as above:

```
systemd.defaultUnit = lib.mkIf (!xcfg.autorun && (xcfg.autorun || dmcfg.enable)) "graphical.target";
```

but that needs a clear comment otherwise we forget why we did that.

---

I was against adding an in tree mkForce to the defaultUnit as that probably causes quite some confusion https://github.com/NixOS/nixpkgs/issues/311683#issuecomment-2523273247

Reply:
Yeah, I like that plus a comment.